### PR TITLE
Improve internal types

### PR DESCRIPTION
**Pull request by wagenet**
Improve types for @ember/object

Includes:
  - Object, { action } from '@ember/object'
  - CoreObject from '@ember/object/core'
  - Observable from '@ember/object/observable'
  - { dependentKeyCompat } from '@ember/object/compat'

Note that types for `extend`, `reopen`, and `reopenClass` are not
exported. Also, `get` and `set` have naive types. All of these will
be deprecated in future releases.

Many types were copied from DT. We should make sure to give appropriate credit, but I suspect we may already be doing that.

### CODE REVIEW

File: packages/@ember/-internals/routing/lib/system/route.ts [hunk 1]
```diff
@@ -2752,16 +2803,7 @@ if (ROUTER_EVENTS) {
     },
   };
 
-  Route.reopen(ROUTER_EVENT_DEPRECATIONS, {
-    _paramsFor(routeName: string, params: {}) {
```
Comment (wagenet) [inline packages/@ember/-internals/routing/lib/system/route.ts:2756]: I'm pretty sure it was a bug for this to be in a conditional reopen. We always depend on `_paramsFor`.

File: packages/@ember/-internals/runtime/types/observable.d.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
+import { ObserverMethod } from './-private/types';
+
+/**
+ * This mixin provides properties and property observing functionality, core features of the Ember object model.
+ */
```
Comment (chriskrycho) [inline packages/@ember/-internals/runtime/types/observable.d.ts:5]: A discussion point (cc. @pzuraq @dfreeman @jamescdavis): throughout, this uses the `keyof this` to make keys support both auto-complete and type-checking against property keys. However, it *drops* the feature from DT where it returns the type. I *think*, if we drop support for building types mixins, we can actually enable `strictFunctionTypes` and keep the return type correct?

File: packages/@ember/object/type-tests/compat.test.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+import { dependentKeyCompat } from '@ember/object/compat';
+import { expectTypeOf } from 'expect-type';
+
+expectTypeOf(dependentKeyCompat).toMatchTypeOf<Function>();
```
Comment (chriskrycho) [inline packages/@ember/object/type-tests/compat.test.ts:4]:
```suggestion
expectTypeOf(dependentKeyCompat).toMatchTypeOf<PropertyDecorator>();
```

**Review by chriskrycho — COMMENTED** [review]
Thanks for tackling this! One note: *if possible* I think it'd be extremely helpful to separate the functional changes from the type-only changes. In particular, I'm extremely leery of any shuffling in the router here (once bitten twice shy, and the router has bitten about four months off of my life 😂).

File: packages/@ember/-internals/routing/lib/system/route.ts [hunk 4]
```diff
@@ -127,6 +130,13 @@ class Route extends EmberObject implements IRoute {
     }
   }
 
+  // Implement Evented
+  on!: (name: string, method: ((...args: any[]) => void) | string) => this;
+  one!: (name: string, method: string | ((...args: any[]) => void)) => this;
+  trigger!: (name: string, ...args: any[]) => any;
```
[the remark this answers is not in the document]
Comment (wagenet) [reply packages/@ember/-internals/routing/lib/system/route.ts:136]: I copied this from evented.d.ts :/

**Review by chriskrycho — APPROVED** [review]
Thanks for these!

File: .eslintrc.js [hunk 5]
```diff
@@ -35,7 +35,7 @@ module.exports = {
 
   overrides: [
     {
-      files: ['*.ts'],
+      files: ['*.ts', '*.d.ts'],
```
[the remark this answers is not in the document]
Comment (wagenet) [reply .eslintrc.js:38]: Yes, it should. I think this must have been from an attempt to fix something I fixed in another way.

[the remark this answers is not in the document]
Comment (wagenet) [reply .eslintrc.js]: Fixed.

File: packages/@ember/object/index.d.ts [hunk 6]
```diff
@@ -1 +1,17 @@
...
+//   defineProperty,
+//   get,
+//   getProperties,
+//   getWithDefault,
+//   notifyPropertyChange,
+//   observer,
+//   set,
+//   setProperties,
+//   trySet,
+// } from '@ember/-internals/metal';
```
Comment (rwjblue) [inline packages/@ember/object/index.d.ts:15]: Why is this here? In general, I think we'd prefer to avoid landing a commented out section like this (basically we should bring it back, delete it, or add a comment explaining why its commented).

**Review by rwjblue — APPROVED** [review]

File: packages/@ember/object/index.d.ts (hunk 6 above)
Comment (wagenet) [reply packages/@ember/object/index.d.ts:15]: It's stuff that we want to export but isn't ready yet. I'll remove it and add it back properly in another PR.

**Comment by stefanpenner** [comment]
@wagenet these files appear to unexpectedly end up in the build output, which seems unintentional.

**Comment by stefanpenner** [comment]
^^ fix landed, all should be resolved.
